Make jar files be osgi bundles

steinarb opened the pull request:
This PR fixes https://github.com/haraldk/TwelveMonkeys/issues/794

With this change, all jar files also becomes OSGi bundles. 

This will not affect non-OSGi applications in any way. It is just some extra headers added to the MANIFEST.MF files.

The commons jar files with get Import-Package and Export-Package MANIFEST.MF headers.

The imageio jar files will get Import-Package and Export-Package MANIFEST.MF headers and in addition Require-Capability headers requiring the osgi.serviceloader.registrar capability and Provide-Capability exporting the SPI services defined in META-INF/services of the imageio jar files.

The changes are:

1. TwelveMonkeys/pom.xml add basic configuration for the apache felix maven-bundle-plugin in the &lt;pluginManagement> section
2. TwelveMonkeys/common/\*/pom.xml reference the maven-bundle-plugin in the &lt;plugins> section
3. TwelveMonkeys/imageio/pom.xml add a Require-Capability header for osgi.serviceloader.registrar to the maven-bundle-plugin config in the &lt;pluginManagement> section
4. TwelveMonkeys/imageio/\*/pom.xml reference the maven-bundle-plugin in the &lt;plugins> section and add a Provide-Capability header exporting the META-INF/services of the jar files with the osgi.serviceloader

haraldk commented:
Thanks! I did a quick review, will perform a full one after the CI has run.

haraldk commented:
...when I have time. Sorry, will look at this, but have been quite busy.

steinarb commented:
>>>>> Harald Kuhr ***@***.***>:

> ...when I have time. Sorry, will look at this, but have been quite busy.

No worries.

I saw there was a CI error for Java 11, but I didn't look like it was
caused by the OSGi manifest changes (at least: I can't think of any way
the manifest changes could be the cause).

steinarb commented on imageio/imageio-jpeg-jai-interop/pom.xml:
```diff
@@ -25,6 +25,10 @@
                     <skip>true</skip>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+            </plugin>
```
Ok, I'll remove,

steinarb replied:
removed

steinarb commented on imageio/imageio-reference/pom.xml:
```diff
@@ -25,6 +25,10 @@
                     <skip>true</skip>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+            </plugin>
```
Ok, will remove

steinarb commented on imageio/imageio-tiff-jai-interop/pom.xml:
```diff
@@ -25,6 +25,10 @@
                     <skip>true</skip>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+            </plugin>
```
will remove

steinarb commented on imageio/imageio-tiff-jdk-interop/pom.xml:
```diff
@@ -25,6 +25,10 @@
                     <skip>true</skip>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+            </plugin>
```
will-remove

steinarb commented on imageio/pom.xml:
```diff
@@ -170,4 +170,22 @@
+                        </instructions>
+                    </configuration>
+                </plugin>
+        </plugins>
```
Will fix.

steinarb commented on pom.xml:
```diff
@@ -218,6 +218,43 @@
+                        <supportedProjectTypes>
+                            <supportedProjectType>jar</supportedProjectType>
+                            <supportedProjectType>bundle</supportedProjectType>
+                            <supportedProjectType>war</supportedProjectType>
```
It's a copy-paste from my own config, here: https://github.com/steinarb/bang-bompom/blob/master/bang-pom/pom.xml#L74

I think I dropped using .war files back in 2017, but have never bothered to remove that. 

But I can remove it here without any side effects so I'll do it.

steinarb commented:
> Mostly questions due to my lack of OSGi experience...
> 
> Would also prefer if you could remove the bundle-plugin in the modules that don't need it, unless there is a specific need for it (in that case, please explain, and I might accept it 😀 ).

Nope I just figured everything under imageio was plugins (even though not all of them had a service).

Will remove.

haraldk commented on .gitignore:
```diff
@@ -15,3 +15,4 @@ private
 profiles.xml
 Thumbs.db
 .DS_Store
+/.metadata/
```
I don't mind having it there. It's a little random in this PR, but just keep it.

steinarb commented:
A fix has been pushed, rebased on the latest TwelveMonkeys master.

I tried to, but was unable to, resolve the conversations in the comments.

Changes:

1. removed maven-bundle-plugin from the following poms:
    1. imageio/imageio-jpeg-jai-interop/pom.xml
    2. imageio/imageio-jpeg-jep262-interop/pom.xml
    3. imageio/imageio-reference/pom.xml
    4. imageio/imageio-tiff-jai-interop/pom.xml
    5. imageio/imageio-tiff-jdk-interop/pom.xml
2. I have fixed formatting in imageio/pom.xml
3. I have removed packaging war from the maven-bundle-config in pom.xml

haraldk reviewed (approved):
Looking good! Thanks!
